lib/reek/smells/instance_variable_assumption.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+require_relative 'smell_detector'
+require_relative 'smell_warning'
+
+module Reek
+  module Smells
+    #
+    # The +InstanceVariableAssumption+ class is responsible for
+    # detecting directly access of instance variables in a class
+    # that does not define them in its initialize method.
+    #
+    class InstanceVariableAssumption < SmellDetector
+      def self.contexts
+        [:class]
+      end
+
+      # Checks +klass+ for instance
+      # variables assumptions.
+      #
+      # @return [Array<SmellWarning>]
+      #
+      def inspect(ctx)
+        ivar_assumptions = class_ivars(ctx) - initialized_ivars(ctx)
+
+        return [] if ivar_assumptions.empty?
+
+        build_smell_warning(ctx, ivar_assumptions.uniq)
+      end
+
+      private
+
+      def build_smell_warning(ctx, ivar_assumptions)
```

Comment:
That's what `smell_warning` was meant for. Perhaps we can indeed extract some common code.